File: helper-code/tack-and-strat-java/src/main/java/com/tackstrat/model/City.java
package com.tackstrat.model;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.EnumSet;
import java.util.Iterator;
import java.util.List;
import java.util.Optional;

/** A founded city: produces +2 production / turn, builds units, has HP and can be captured. */
public final class City {

    public static final int PRODUCTION_PER_TURN = 2;
    public static final int MAX_HP = 20;

    private final int id;
    private int ownerSeat;
    private final HexCoord coord;
    private final String name;
    private UnitKind currentBuild;
    private final Deque<UnitKind> buildQueue = new ArrayDeque<>();
    private int productionStored;
    private int population;
    private int foodStored;
    private int hp;
    private CityFocus focus = CityFocus.BALANCED;
    private final EnumSet<CityBuilding> buildings = EnumSet.noneOf(CityBuilding.class);
    /** Active hunting parties: population was reduced when each started. */
    private final ArrayList<HuntMission> huntMissions = new ArrayList<>();

    public static final class HuntMission {
        public int turnsRemaining;
        public final int targetAnimalId;

        public HuntMission(int turnsRemaining, int targetAnimalId) {
            this.turnsRemaining = turnsRemaining;
            this.targetAnimalId = targetAnimalId;
        }
    }

    public City(int id, int ownerSeat, HexCoord coord, String name, UnitKind initialBuild) {
        this.id = id;
        this.ownerSeat = ownerSeat;
        this.coord = coord;
        this.name = name;
        this.currentBuild = initialBuild;
        this.productionStored = 0;
        this.population = 1;
        this.foodStored = 0;
        this.hp = MAX_HP;
    }

    public int id() {
        return id;
    }

    public int ownerSeat() {
        return ownerSeat;
    }

    public HexCoord coord() {
        return coord;
    }

    public String name() {
        return name;
    }

    public UnitKind currentBuild() {
        return currentBuild;
    }

    public int productionStored() {
        return productionStored;
    }

    public int hp() {
        return hp;
    }

    public int population() {
        return population;
    }

    public int foodStored() {
        return foodStored;
    }

    public CityFocus focus() {
        return focus;
    }

    public boolean hasBuilding(CityBuilding b) {
        return buildings.contains(b);
    }

    public EnumSet<CityBuilding> buildings() {
        return EnumSet.copyOf(buildings);
    }

    /** Future builds after {@link #currentBuild()}. */
    public List<UnitKind> queuedBuilds() {
        return List.copyOf(buildQueue);
    }

    public int maxHp() {
        return MAX_HP;
    }

    public List<HuntMission> huntMissions() {
        return List.copyOf(huntMissions);
    }

    public int huntPartiesAway() {
        return huntMissions.size();
    }

    /**
     * Send a townsfolk hunting party. Population drops by 1 until the party returns.
     *
     * @return empty if started, or error message
     */
    public Optional<String> startHuntMission(int targetAnimalId, int roundsAway) {
        if (population < 2) {
            return Optional.of("Need at least 2 population to send a hunter.");
        }
        if (roundsAway < 1) {
            return Optional.of("Invalid hunt length.");
        }
        population--;
        huntMissions.add(new HuntMission(roundsAway, targetAnimalId));
        return Optional.empty();
    }

    /** End-of-turn tick: returning hunters restore +1 pop and complete the hunt against the target animal. */
    public List<Integer> advanceHuntMissions() {
        List<Integer> animalsToCull = new ArrayList<>();
        Iterator<HuntMission> it = huntMissions.iterator();
        while (it.hasNext()) {
            HuntMission m = it.next();
            m.turnsRemaining--;
            if (m.turnsRemaining <= 0) {
                population++;
                animalsToCull.add(m.targetAnimalId);
                it.remove();
            }
        }
        return animalsToCull;
    }

    /** Animals cannot wipe a city; HP is floored so it always remains contestable by units. */
    public void applyWildlifeRaid(int dmg) {
        hp = Math.max(1, hp - Math.max(0, dmg));
    }

    public void setCurrentBuild(UnitKind kind) {
        if (kind != currentBuild) {
            currentBuild = kind;
            productionStored = 0;
        }
    }

    public void addProduction(int amount) {
        productionStored += amount;
    }

    public void setFocus(CityFocus focus) {
        this.focus = focus == null ? CityFocus.BALANCED : focus;
    }

    public boolean addBuilding(CityBuilding building) {
        if (building == null) return false;
        return buildings.add(building);
    }

    public void applyFoodYield(int foodYield) {
        int upkeep = 1 + population;
        int surplus = foodYield - upkeep;
        if (surplus <= 0) {
            foodStored = Math.max(0, foodStored + surplus);
            return;
        }
        foodStored += surplus;
        while (foodStored >= growthThreshold()) {
            foodStored -= growthThreshold();
            population++;
        }
    }

    /** Foraging return from a hunting party (no upkeep subtraction). */
    public void addForagedFood(int food) {
        if (food <= 0) return;
        foodStored += food;
        while (foodStored >= growthThreshold()) {
            foodStored -= growthThreshold();
            population++;
        }
    }

    /** Reserve one population to form an external party; leaves at least 1 in city. */
    public boolean consumePopulationForParty() {
        if (population < 2) return false;
        population--;
        return true;
    }

    /** Party returns home and becomes population again. */
    public void returnPopulationFromParty() {
        population++;
    }

    public int growthThreshold() {
        int base = 8 + population * 4;
        if (hasBuilding(CityBuilding.GRANARY)) {
            base = Math.max(4, base - 2);
        }
        return base;
    }

    public void enqueueBuild(UnitKind kind) {
        if (currentBuild == null) {
            currentBuild = kind;
            return;
        }
        buildQueue.addLast(kind);
    }

    /** Remove queued item at index (0-based). */
    public boolean removeQueuedBuild(int index) {
        if (index < 0 || index >= buildQueue.size()) return false;
        var list = new ArrayList<>(buildQueue);
        list.remove(index);
        buildQueue.clear();
        buildQueue.addAll(list);
        return true;
    }

    /** Move queued item one position up/down; delta should be -1 or +1. */
    public boolean moveQueuedBuild(int index, int delta) {
        if (index < 0 || index >= buildQueue.size()) return false;
        int to = index + delta;
        if (to < 0 || to >= buildQueue.size()) return false;
        var list = new ArrayList<>(buildQueue);
        var item = list.remove(index);
        list.add(to, item);
        buildQueue.clear();
        buildQueue.addAll(list);
        return true;
    }

    public java.util.Optional<UnitKind> drainCompleted() {
        if (currentBuild == null) {
            return java.util.Optional.empty();
        }
        int cost = currentBuild.productionCost();
        if (productionStored < cost) {
            return java.util.Optional.empty();
        }
        UnitKind produced = currentBuild;
        productionStored -= cost;
        if (!buildQueue.isEmpty()) {
            currentBuild = buildQueue.removeFirst();
        } else {
            // Require explicit next choice; do not auto-repeat previous build forever.
            currentBuild = null;
        }
        return java.util.Optional.of(produced);
    }

    /** Hand the city to a new owner; reset HP so it isn't immediately recapturable. */
    public void captureBy(int newOwnerSeat) {
        ownerSeat = newOwnerSeat;
        hp = MAX_HP / 2;
        productionStored = 0;
    }

    /** Restore mutable fields when loading a save (bypasses {@link #setCurrentBuild} side effects). */
    public void applySavedState(int savedOwnerSeat, UnitKind savedBuild, int savedProductionStored, int savedHp) {
        applySavedState(savedOwnerSeat, savedBuild, List.of(), savedProductionStored, savedHp, 1, 0, null);
    }

    /** Restore mutable fields when loading a save, including queued builds. */
    public void applySavedState(
            int savedOwnerSeat,
            UnitKind savedBuild,
            List<UnitKind> savedQueue,
            int savedProductionStored,
            int savedHp,
            int savedPopulation,
            int savedFoodStored) {
        applySavedState(
                savedOwnerSeat, savedBuild, savedQueue, savedProductionStored, savedHp, savedPopulation, savedFoodStored,
                null);
    }

    public void applySavedState(
            int savedOwnerSeat,
            UnitKind savedBuild,
            List<UnitKind> savedQueue,
            int savedProductionStored,
            int savedHp,
            int savedPopulation,
            int savedFoodStored,
            List<HuntMission> savedHunts) {
        if (savedHp < 0 || savedHp > MAX_HP) {
            throw new IllegalArgumentException("Invalid city HP: " + savedHp);
        }
        if (savedProductionStored < 0) {
            throw new IllegalArgumentException("Invalid production: " + savedProductionStored);
        }
        if (savedPopulation < 1) {
            throw new IllegalArgumentException("Invalid population: " + savedPopulation);
        }
        if (savedFoodStored < 0) {
            throw new IllegalArgumentException("Invalid food storage: " + savedFoodStored);
        }
        this.ownerSeat = savedOwnerSeat;
        this.currentBuild = savedBuild;
        this.buildQueue.clear();
        this.buildQueue.addAll(new ArrayList<>(savedQueue));
        this.productionStored = savedProductionStored;
        this.population = savedPopulation;
        this.foodStored = savedFoodStored;
        this.hp = savedHp;
        this.focus = CityFocus.BALANCED;
        this.buildings.clear();
        huntMissions.clear();
        if (savedHunts != null) {
            huntMissions.addAll(savedHunts);
        }
    }

    public void applySavedCityMeta(CityFocus savedFocus, List<CityBuilding> savedBuildings) {
        this.focus = savedFocus == null ? CityFocus.BALANCED : savedFocus;
        this.buildings.clear();
        if (savedBuildings != null) {
            this.buildings.addAll(savedBuildings);
        }
    }
}

File: src/engine/city.cpp
#include "tack/strat/city.hpp"

#include <algorithm>
#include <stdexcept>

namespace tack::strat {

City::City(int id, int owner_seat, HexCoord coord, std::string name,
           std::optional<UnitKind> initial_build)
    : id_(id),
      owner_seat_(owner_seat),
      coord_(coord),
      name_(std::move(name)),
      current_build_(initial_build) {}

std::vector<UnitKind> City::queued_builds() const {
  return std::vector<UnitKind>(build_queue_.begin(), build_queue_.end());
}

std::optional<std::string> City::start_hunt_mission(int target_animal_id, int rounds_away) {
  if (population_ < 2) {
    return std::string("Need at least 2 population to send a hunter.");
  }
  if (rounds_away < 1) {
    return std::string("Invalid hunt length.");
  }
  population_--;
  hunt_missions_.push_back(HuntMission{rounds_away, target_animal_id});
  return std::nullopt;
}

std::vector<int> City::advance_hunt_missions() {
  std::vector<int> animals_to_cull;
  for (auto it = hunt_missions_.begin(); it != hunt_missions_.end();) {
    it->turns_remaining--;
    if (it->turns_remaining <= 0) {
      population_++;
      animals_to_cull.push_back(it->target_animal_id);
      it = hunt_missions_.erase(it);
    } else {
      ++it;
    }
  }
  return animals_to_cull;
}

void City::apply_wildlife_raid(int dmg) noexcept {
  hp_ = std::max(1, hp_ - std::max(0, dmg));
}

void City::set_current_build(std::optional<UnitKind> kind) {
  if (kind != current_build_) {
    current_build_ = kind;
    production_stored_ = 0;
  }
}

bool City::add_building(CityBuilding building) {
  auto [_, inserted] = buildings_.insert(building);
  return inserted;
}

void City::apply_food_yield(int food_yield) {
  int upkeep = 1 + population_;
  int surplus = food_yield - upkeep;
  if (surplus <= 0) {
    food_stored_ = std::max(0, food_stored_ + surplus);
    return;
  }
  food_stored_ += surplus;
  while (food_stored_ >= growth_threshold()) {
    food_stored_ -= growth_threshold();
    population_++;
  }
}

void City::add_foraged_food(int food) {
  if (food <= 0) {
    return;
  }
  food_stored_ += food;
  while (food_stored_ >= growth_threshold()) {
    food_stored_ -= growth_threshold();
    population_++;
  }
}

bool City::consume_population_for_party() {
  if (population_ < 2) {
    return false;
  }
  population_--;
  return true;
}

int City::growth_threshold() const {
  int base = 8 + population_ * 4;
  if (has_building(CityBuilding::GRANARY)) {
    base = std::max(4, base - 2);
  }
  return base;
}

void City::enqueue_build(UnitKind kind) {
  if (!current_build_.has_value()) {
    current_build_ = kind;
    return;
  }
  build_queue_.push_back(kind);
}

bool City::remove_queued_build(int index) {
  if (index < 0 || index >= static_cast<int>(build_queue_.size())) {
    return false;
  }
  auto it = build_queue_.begin();
  std::advance(it, index);
  build_queue_.erase(it);
  return true;
}

bool City::move_queued_build(int index, int delta) {
  if (index < 0 || index >= static_cast<int>(build_queue_.size())) {
    return false;
  }
  int const to = index + delta;
  if (to < 0 || to >= static_cast<int>(build_queue_.size())) {
    return false;
  }
  std::vector<UnitKind> list(build_queue_.begin(), build_queue_.end());
  UnitKind item = list[static_cast<std::size_t>(index)];
  list.erase(list.begin() + index);
  list.insert(list.begin() + to, item);
  build_queue_.clear();
  for (UnitKind k : list) {
    build_queue_.push_back(k);
  }
  return true;
}

std::optional<UnitKind> City::drain_completed() {
  if (!current_build_.has_value()) {
    return std::nullopt;
  }
  int cost = unit_production_cost(*current_build_);
  if (production_stored_ < cost) {
    return std::nullopt;
  }
  UnitKind produced = *current_build_;
  production_stored_ -= cost;
  if (!build_queue_.empty()) {
    current_build_ = build_queue_.front();
    build_queue_.pop_front();
  } else {
    current_build_ = std::nullopt;
  }
  return produced;
}

void City::capture_by(int new_owner_seat) noexcept {
  owner_seat_ = new_owner_seat;
  hp_ = MAX_HP / 2;
  production_stored_ = 0;
}

void City::apply_saved_state(int saved_owner_seat, std::optional<UnitKind> saved_build,
                             std::vector<UnitKind> saved_queue, int saved_production_stored,
                             int saved_hp, int saved_population, int saved_food_stored,
                             std::vector<HuntMission> saved_hunts) {
  if (saved_hp < 0 || saved_hp > MAX_HP) {
    throw std::invalid_argument("Invalid city HP");
  }
  if (saved_production_stored < 0) {
    throw std::invalid_argument("Invalid production");
  }
  if (saved_population < 1) {
    throw std::invalid_argument("Invalid population");
  }
  if (saved_food_stored < 0) {
    throw std::invalid_argument("Invalid food storage");
  }
  owner_seat_ = saved_owner_seat;
  current_build_ = saved_build;
  build_queue_.clear();
  for (UnitKind k : saved_queue) {
    build_queue_.push_back(k);
  }
  production_stored_ = saved_production_stored;
  population_ = saved_population;
  food_stored_ = saved_food_stored;
  hp_ = saved_hp;
  focus_ = CityFocus::BALANCED;
  buildings_.clear();
  hunt_missions_ = std::move(saved_hunts);
}

void City::apply_saved_city_meta(CityFocus saved_focus, std::vector<CityBuilding> saved_buildings) {
  focus_ = saved_focus;
  buildings_.clear();
  for (CityBuilding b : saved_buildings) {
    buildings_.insert(b);
  }
}

}  // namespace tack::strat
